document-portal/document-portal.c

```diff
@@ -431,85 +433,43 @@ portal_add (GDBusMethodInvocation *invocation,
             GVariant              *parameters,
             XdpAppInfo            *app_info)
 {
+  int fd_id;
+  gboolean reuse_existing, persistent;
+  DocumentAddFullFlags flags = 0;
   GDBusMessage *message;
   GUnixFDList *fd_list;
-  g_autofree char *id = NULL;
-  int fd_id, fd, fds_len;
-  g_autofree char *path = NULL;
-  gboolean writable;
-  const int *fds;
-  struct stat st_buf, real_parent_st_buf;
-  gboolean reuse_existing, persistent;
   GError *error = NULL;
-  const char *app_id = xdp_app_info_get_id (app_info);
+  g_auto(GStrv) ids = NULL;
 
   g_variant_get (parameters, "(hbb)", &fd_id, &reuse_existing, &persistent);
 
+  if (reuse_existing)
+    flags |= DOCUMENT_ADD_FLAGS_REUSE_EXISTING;
+  if (persistent)
+    flags |= DOCUMENT_ADD_FLAGS_PERSISTENT;
+ 
   message = g_dbus_method_invocation_get_message (invocation);
   fd_list = g_dbus_message_get_unix_fd_list (message);
 
-  fd = -1;
   if (fd_list != NULL)
     {
-      fds = g_unix_fd_list_peek_fds (fd_list, &fds_len);
+      int fds_len;
+      const int *fds = g_unix_fd_list_peek_fds (fd_list, &fds_len);
       if (fd_id < fds_len)
-        fd = fds[fd_id];
+        {
+          int fd = fds[fd_id];
+
+          ids = document_add_full (&fd, 1, flags, app_info, "", 0, &error);
```

Comment:
This seems to change the behaviour wrt write access. Before it gave the app write access iff the passed in fd was writable. However, now it never gets it (because we pass 0 as target_perms).

Reply:
Ah, no because that is target perms, not caller perms.